objcxx_eh.cc

```diff
@@ -515,14 +519,71 @@ static void eh_cleanup(void *exception)
 	objc_release(*(id*)exception);
 }
 
+namespace __cxxabiv1
+{
+    struct __cxa_exception
+    {
+		// Manage the exception object itself.
+		std::type_info *exceptionType;
+		void (*exceptionDestructor)(void *);
+		terminate_handler unexpectedHandler;
+		terminate_handler terminateHandler;
+		__cxa_exception *nextException;
+		int handlerCount;
+		int handlerSwitchValue;
+		const unsigned char *actionRecord;
+		const unsigned char *languageSpecificData;
+		void *catchTemp;
+		void *adjustedPtr;
+		_Unwind_Exception unwindHeader;
+    };
+
+	struct __cxa_refcounted_exception
+    {
+        // Manage this header.
+        int referenceCount;
+        // __cxa_exception must be last, and no padding can be after it.
+        __cxa_exception exc;
+    };
+}
+
+extern "C" __cxa_refcounted_exception* __cxa_init_primary_exception(void *obj, std::type_info *tinfo, void (*dest) (void *));
+
 extern "C"
 OBJC_PUBLIC
 void objc_exception_throw(id object)
 {
-	id *exc = (id *)__cxa_allocate_exception(sizeof(id));
-	*exc = object;
-	objc_retain(object);
-	DEBUG_LOG("objc_exception_throw: Throwing 0x%x\n", *exc);
+    id *exc = (id *)__cxa_allocate_exception(sizeof(id));
+    *exc = object;
+    objc_retain(object);
+    DEBUG_LOG("objc_exception_throw: Throwing 0x%x\n", *exc);
+
+#ifdef _LIBCPP_VERSION
```

Comment:
This is reintroducing tight coupling with C++ runtime implementations, which we had previously removed.  It's also totally the wrong check, because you're checking for libc++ but libc++ can be built with libc++abi, libcxxrt, or libsupc++.

Reply:
This `#ifdef` is trying to determine whether you're on the mingw64 or ucrt64 MSYS2 environment (which use libstdc++ and have `__cxa_init_primary_exception`) or the clang64 MSYS2 environment, which doesn't have `__cxa_init_primary_exception` and for which we revert to the old behavior of using `__cxa_throw`. 

I've redefined the check as `#ifndef __GLIBCXX__`, but not sure if that really addresses your comment.

As for the tight coupling: yes, I know, though we tried the dynamic check for the exception struct size and that didn't work.

Reply:
Hmm -- that didn't work (broke the clang64 build), so sticking with `#ifdef _LIBCPP_VERSION` for now